sled-agent/src/hardware/illumos/mod.rs

```diff
@@ -40,6 +40,28 @@ enum Error {
     NoDevLinks(PathBuf),
 }
 
+const GIMLET_ROOT_NODE_NAME: &str = "Oxide,Gimlet";
+
+/// Return true if the host system is an Oxide Gimlet.
+#[cfg(not(test))]
+pub fn is_gimlet() -> anyhow::Result<bool> {
+    let mut device_info = DevInfo::new_force_load()?;
+    let mut node_walker = device_info.walk_node();
+    let Some(root) = node_walker.next().transpose()? else {
+        return Err(anyhow::anyhow!("No nodes in device tree"));
+    };
+    Ok(root.node_name() == GIMLET_ROOT_NODE_NAME)
+}
+
+/// Return true if the host system is an Oxide Gimlet.
+//
+// TODO-testing: This assumes we never test on real Gimlets. That seems like a
+// very bad assumption.
```

Comment:
Agreed :). A couple suggestions:

* Can the permission problem be fixed in CI? (Do we need to `pfexec` the test runs?)
* If not, as a slightly-less-painful bandaid could we change this to call all the same functions that `is_gimlet` calls, but return `Ok(false)` on any error? I.e., the `#[cfg(test)]` version never fails, but on a real gimlet with correct permissions it could still return `Ok(true)`.

Reply:
>  Can the permission problem be fixed in CI? (Do we need to pfexec the test runs?)

I'm not sure, though I'm a bit hesitant to `pfexec` the _entire_ test suite. It's all in a VM, I guess, but it feels like a big hammer to me.

> If not, as a slightly-less-painful bandaid could we change this to call all the same functions that is_gimlet calls, but return Ok(false) on any error? I.e., the #[cfg(test)] version never fails, but on a real gimlet with correct permissions it could still return Ok(true).

That might make sense, yeah. We're currently running this in three contexts: non-Gimlet developer machines, CI, and potentially real Gimlets. In the first two, `Ok(false)` is _correct_ even if we get the answer for the wrong reason. On the latter, we'll get the right answer since everyone is currently logging in as `root` anyway. That's not great, but it's a separate issue from this one. When we deal with it, we can be sure to configure permissions such that calling `libdevinfo` doesn't fail anyway.

Thanks for the suggestion @jgallagher, very helpful as usual.

Reply:
For what it's worth if you're having devinfo permissions errors it's probably because we're passing a flag to try and force attach stuff, which requires root.  It would be reasonable at least for the "am I a Gimlet" case to _not_ pass that flag, as I believe you're just looking at properties on the root node which is always there and doesn't require any flags to see.

Reply:
Thanks Josh, that's helpful. I think we _are_ indeed force-attaching. I'll try it out with out that now, and hopefully we can remove the test-only version of this.